### PR TITLE
fix(headers): require Akamai body fingerprint, not just edge Server header

### LINKED ISSUE #194: Bug: Akamai WAF-block detection false-positives on ordinary origin error responses
## Summary

`_detect_waf_block_page`'s Akamai signature — `("server", "akamaighost", "Akamai", 400)` — classifies a response as an Akamai WAF challenge/block page solely from the presence of `Server: AkamaiGHost` combined with `status_code >= 400`.

Neither signal is sufficiently specific to a bot-mitigation event. `Server: AkamaiGHost` identifies Akamai's edge infrastructure; it does not indicate that the current request was blocked. Likewise, `status_code >= 400` covers ordinary HTTP errors such as `404`, `401`, `403`, `500`, `502`, and `503`.

As a result, an ordinary error response from an Akamai-fronted site can be incorrectly classified as a bot-detection block, causing `headers_analyzer` to abort header analysis and return a misleading error to the caller.

## Current implementation

```python
_WAF_BLOCK_PAGE_HEADER_SIGNATURES = (
    ("cf-mitigated", "challenge", "Cloudflare", None),
    ("server", "akamaighost", "Akamai", 400),
)

def _detect_waf_block_page(raw_headers: dict, status_code: int, body: str) -> str | None:
    for header_name, expected_value, provider, min_status in _WAF_BLOCK_PAGE_HEADER_SIGNATURES:
        if raw_headers.get(header_name, "").lower() != expected_value:
            continue
        if min_status is not None and status_code < min_status:
            continue
        return provider
    ...
```

`headers_analyzer` then short-circuits when a provider is returned:

```python
if blocked_by:
    return {
        "success": False,
        "error": (
            f"Received a bot-detection challenge/block page instead "
            f"of the real site ({blocked_by} detected). ..."
        ),
    }
```

## Root cause

The Akamai signature treats CDN identification as equivalent to WAF mitigation:

```text
Server: AkamaiGHost
        +
status_code >= 400
        ↓
Akamai block
```

However, an Akamai-fronted site can legitimately return a 4xx/5xx response without Akamai blocking the request. The current signature has no signal that distinguishes an actual Akamai mitigation response from an ordinary error response passed through the Akamai edge.

This is the same class of problem already accounted for elsewhere in this file for Imperva/Incapsula. The existing implementation recognizes that headers such as `X-Iinfo`/`X-CDN` can occur on both normal and blocked responses and therefore requires a distinctive body fingerprint (`_incapsula_resource`) as an additional signal.

The Akamai signature does not currently have an equivalent secondary fingerprint and instead relies on the generic `status_code >= 400` condition, which is insufficient to establish that a WAF block occurred.

## Reproduction

The public `headers_analyzer` API accepts a domain/FQDN rather than an arbitrary URL path, so this issue does **not** require a nonexistent URL path such as `/does-not-exist`.

The false positive can be reproduced directly at the detection layer with an ordinary error response:

```python
def test_akamai_ordinary_404_is_not_block_page(self):
    result = _detect_waf_block_page(
        {"server": "akamaighost"},
        404,
        "<html><body>404 Not Found</body></html>",
    )

    self.assertIsNone(result)
```

With the current implementation, this returns:

```python
"Akamai"
```

because:

```text
server.lower() == "akamaighost"   → True
404 >= 400                        → True
```

The same false classification occurs for ordinary `401`, `403`, `500`, `502`, and `503` responses when the `Server: AkamaiGHost` header is present.

A live reproduction through `headers_analyzer` should use a valid Akamai-fronted domain whose root URL itself returns an ordinary 4xx/5xx response. The domain validation performed by `is_valid_domain()` is unrelated to this issue because it validates the input domain, while this bug concerns classification of the subsequent HTTP response.

## Impact

- This is not merely a mis-tagged severity. The tool **aborts header analysis entirely** and reports a fabricated bot-detection cause for an ordinary HTTP error.
- Legitimate security-header findings can be hidden because `headers_analyzer` returns before performing the normal analysis.
- The false positive can affect any Akamai-fronted target whose response happens to satisfy the current `Server` + `>=400` signature, including ordinary authorization failures, missing resources, and server-side errors.

## Proposed fix direction

Strengthen the Akamai detection logic so that `Server: AkamaiGHost` combined with a generic 4xx/5xx status is **not sufficient by itself** to classify a response as a WAF block.

The implementation should require an additional, distinctive signal from an actual Akamai block/challenge response. A body-content fingerprint is one possible approach and would align with the existing Imperva/Incapsula detection pattern.

Akamai block responses may contain distinctive markers such as an `Access Denied` message and a `Reference #...` identifier, but the exact fingerprint should be verified against a captured/confirmed Akamai block response before being added to the implementation.

The fix should preserve detection of genuine Akamai block/challenge pages while allowing ordinary Akamai-fronted 4xx/5xx responses to proceed through normal header analysis.

## Test coverage required

The fix should include regression tests covering at minimum:

- `Server: AkamaiGHost` + `404` + ordinary 404 body → **must not** be classified as `Akamai`-blocked.
- `Server: AkamaiGHost` + `403` + ordinary `"Forbidden"` body → **must not** be classified as blocked.
- `Server: AkamaiGHost` + `500`/`502`/`503` + generic error body → **must not** be classified as blocked.
- `Server: AkamaiGHost` + appropriate block-page status + verified Akamai block-page fingerprint → **should** be classified as `Akamai`-blocked.
- `Server: AkamaiGHost` + `200` + normal body → **must not** be classified as blocked.

The positive test is important to ensure the fix does not simply remove Akamai detection altogether.

## Out of scope

This issue is limited to the Akamai signature in `_detect_waf_block_page`.

It does not cover:

- Cloudflare's `cf-mitigated` signature
- Imperva/Incapsula detection
- General refactoring of the WAF-detection architecture

Please keep changes scoped to the Akamai detection logic and its associated tests.

## Files to be touched

- `tools/headers_tool.py`
- `tests/test_headers_tool.py`

## Contributor coordination

Please comment on this issue before starting work to avoid duplicate effort. If you intend to work on it, wait for maintainer acknowledgment before opening a PR.

### PR DESCRIPTION
**Pull request by v01dst**
## Summary

`_detect_waf_block_page`'s Akamai signature — `Server: AkamaiGHost` + `status_code >= 400` — classified **any** error response from an Akamai-fronted origin as a WAF block, because `AkamaiGHost` is the edge's own identity, not a mitigation signal. Ordinary 401/403/404/5xx responses passed straight through the edge carrying that header, so `headers_analyzer` aborted and reported a fabricated bot-detection cause instead of analyzing (issue #194).

## Fix

Akamai now follows the same body-fingerprint pattern already used for Imperva/Incapsula:

- **Removed** the `("server", "akamaighost", "Akamai", 400)` header signature.
- **Added** a body signature requiring **both** `"access denied"` **and** `"reference #"` (AND semantics) on an error-class status (`>= 400`) — real Akamai deny pages pair those two markers; origin error pages don't.
- The body-signature loop now supports a needle being a `str` (single) or a tuple (ALL must match), so the Imperva path is unchanged.
- `Server: AkamaiGHost` on a 2xx was already not a block; that behavior is preserved and now tested.

## Regression coverage (`tests/test_headers_tool.py`)

Per the issue's required matrix, at the `_detect` level and through `headers_analyzer`:

- `AkamaiGHost` + 404 / 403 / 500 / 502 / 503 + ordinary bodies → **analyzed normally** (no block classification)
- `AkamaiGHost` + block-page status + Access-Denied + Reference-# fingerprint → **classified as Akamai block**
- Fingerprint on a 200 → not blocked (status floor enforced)
- Only one of the two markers → not blocked (AND semantics)
- Case-insensitivity verified (`AkamaiGHost` / uppercase body)
- Existing block-page test updated to carry the fingerprint in its mock body (a real deny page has one)
- Cloudflare `cf-mitigated` and Imperva `_Incapsula_Resource` paths confirmed unchanged

`tools/headers_tool.py` and `tests/test_headers_tool.py` only, as scoped in the issue.

Fixes #194

### CODE REVIEW

File: tests/test_headers_tool.py [hunk 1]
```diff
@@ -549,6 +557,55 @@ def test_normal_akamai_fronted_site_is_analyzed_normally(self, mock_get):
...
+
+    @patch("tools.headers_tool.requests.get")
+    def test_akamai_partial_fingerprint_is_not_blocked(self, mock_get):
+        # Only one of the two body markers -> ordinary content (e.g. a
+        # forum post mentioning "access denied"), not a block page.
+        mock_get.return_value = _resp(403, {
+            "Server": "AkamaiGHost",
+        }, body="<html><body>Access Denied is what the admin said</body></html>")
+        result = headers_analyzer("example.com")
+        self.assertTrue(result["success"])
```
Comment (gaoharimran29-glitch) [inline tests/test_headers_tool.py:613]: This test only covers one direction of the partial-match case: `"access denied"` present, `"reference #"` missing. The AND-check in `_detect_waf_block_page` — `all(needle in body_lower for needle in required)` — has a second, distinct path through it that isn't exercised anywhere in this file: `"reference #"` present without `"access denied"`. That's a plausible real body too, e.g. a support/ticketing page that legitimately cites a reference number for unrelated reasons.

So please add a test like `test_akamai_partial_fingerprint_reference_only_is_not_blocked`

File: tests/test_headers_tool.py [hunk 2]
```diff
@@ -549,6 +557,55 @@ def test_normal_akamai_fronted_site_is_analyzed_normally(self, mock_get):
...
+    @patch("tools.headers_tool.requests.get")
+    def test_akamai_ordinary_404_is_not_classified_as_block(self, mock_get):
+        # #194 regression: Server: AkamaiGHost + 404 + an ordinary origin
+        # error body must NOT abort analysis as a bot-detection block.
+        mock_get.return_value = _resp(404, {
+            "Server": "AkamaiGHost",
+            "X-Frame-Options": "SAMEORIGIN",
+        }, body="<html><body><h1>404 Not Found</h1></body></html>")
+        result = headers_analyzer("example.com")
+        self.assertTrue(result["success"])
```
Comment (gaoharimran29-glitch) [inline tests/test_headers_tool.py:569]: Requesting change here, and this applies as a pattern across every `..._is_not_classified_as_block` test in this file (404, 403, 5xx, the fingerprint-without-status-floor test, and the partial-fingerprint test) — flagging it on this one since it's the first, but asking for the fix across all of them, not just this instance.

Each of these terminates on `self.assertTrue(result["success"])` and stops there. That assertion only proves the WAF-block short-circuit didn't fire — it has zero visibility into whether the header-analysis pipeline actually executed correctly downstream of that. `success == True` would still pass even if a later refactor silently broke the analysis step and returned a `headers` dict that's empty, truncated, or missing keys entirely. None of these tests would catch that regression, because none of them look past the boolean.

**What I want changed:** each of these tests needs to additionally assert against the real, analyzed header output — not just the top-level `success` flag — so the test actually proves the full pipeline ran correctly, not merely that it wasn't rejected. Concretely, for this test:
```python
self.assertTrue(result["success"])
self.assertEqual(result["headers"]["x-frame-options"]["value"], "SAMEORIGIN")
```
since the mock here already sets `X-Frame-Options: SAMEORIGIN` — this costs nothing, the value is already sitting right there in the mock and just isn't being checked.

Please apply the equivalent in every other test following this same `..._is_not_classified_as_block` pattern, asserting on whatever header value each individual mock already sets (e.g. assert on `result["headers"]["server"]["value"] == "AkamaiGHost"` for the mocks that only set `Server` and nothing else). Every one of these tests should end with at least one concrete assertion tied to the mock's actual header content, not just a boolean check.

**Review by gaoharimran29-glitch — CHANGES_REQUESTED** [review]
Requesting changes — see inline comments for the two blocking test gaps (assertion strength across the `..._is_not_classified_as_block` tests, and missing reverse-direction coverage in `test_akamai_partial_fingerprint_is_not_blocked`). Please also resolve Copilot's two review comments in this same PR rather than splitting into a follow-up — both are one-line fixes.

Production fix itself is correct and resolves #194 as intended. Re-review and approve once the inline items and Copilot's comments are addressed.

File: tools/headers_tool.py [hunk 3]
```diff
@@ -64,8 +65,22 @@
 # confirmed live sample, no proof this string is exclusive to block
 # pages, so an error-class status is required as a second signal
```
[the remark this answers is not in the document]
Comment (v01dst) [reply tools/headers_tool.py:66]: Fixed — the Imperva comment no longer points at 'Akamai's entry above'; it now references the policy the Akamai entry below follows.

File: tools/headers_tool.py [hunk 4]
```diff
@@ -84,8 +99,9 @@ def _detect_waf_block_page(raw_headers: dict, status_code: int, body: str) -> st
         return provider
 
     body_lower = body.lower()
-    for needle, provider, min_status in _WAF_BLOCK_PAGE_BODY_SIGNATURES:
-        if needle not in body_lower:
+    for needles, provider, min_status in _WAF_BLOCK_PAGE_BODY_SIGNATURES:
+        required = (needles,) if isinstance(needles, str) else needles
+        if not all(needle in body_lower for needle in required):
             continue
         if min_status is not None and status_code < min_status:
```
[the remark this answers is not in the document]
Comment (v01dst) [reply tools/headers_tool.py]: Fixed — the min-status check now runs before the body substring scans, so out-of-band responses skip the searches entirely.

File: tests/test_headers_tool.py (hunk 1 above)
Comment (v01dst) [reply tests/test_headers_tool.py:613]: Added — test_akamai_partial_fingerprint_reference_only_is_not_blocked covers the reference-only path through the AND-check.

File: tests/test_headers_tool.py (hunk 2 above)
Comment (v01dst) [reply tests/test_headers_tool.py:569]: Done across all five ..._is_not_classified_as_block tests — each now asserts on the analyzed header value its mock already sets (x-frame-options SAMEORIGIN/DENY, or server AkamaiGHost), so a refactor that empties the headers dict can no longer hide behind the success boolean.

**Comment by v01dst** [comment]
Review feedback addressed in 1a4ad4c:

1. **Stale Imperva comment** (Copilot) — reworded to reference "the policy the Akamai entry below follows".
2. **Min-status after substring scan** (Copilot) — hoisted ahead of the scans; 2xx/3xx responses now skip the body searches entirely.
3. **Missing reference-only path** — added `test_akamai_partial_fingerprint_reference_only_is_not_blocked` ("Reference #…" without "Access Denied", e.g. a ticketing page citing a reference number).
4. **Boolean-only assertions** — every `..._is_not_classified_as_block` test now also asserts on the analyzed header output its mock already provides (`x-frame-options` value, or `server` value), so a regression that empties the headers dict can't pass as "analyzed".

Verification: 65/66 tests pass locally — the single failure (`test_dict_headers_items_preserves_genuinely_duplicate_headers`) is a pre-existing artifact of the local curl_cffi stub and fails identically on the unmodified branch. The other 8 akamai tests all pass, including the new one.

**Review by gaoharimran29-glitch — APPROVED** [review]
All blocking items and Copilot's comments addressed. LGTM Approving.

**Comment by gaoharimran29-glitch** [comment]
@v01dst Its merged now. Thanks for working on it. If you found the project useful, please star this repo. It will help the project to gain more visibility
